cmd/mt-kafka-mdm-report-out-of-order/inputooofinder.go

```diff
@@ -0,0 +1,147 @@
+}
+
+func newInputOOOFinder(prefix string, substr string, partitionFrom int, partitionTo int, metricInterval int, reorderWindow int, groupByName bool, groupedByName *map[string]int, groupByTag string, groupedByTag *map[string]int) *inputOOOFinder {
+	cassandraIndex := cassandra.New(cassandra.CliConfig)
+	err := cassandraIndex.InitBare()
+	if err != nil {
+		log.Fatalf("error initializing cassandra index: %s", err.Error())
+		os.Exit(1)
+	}
+
+	metricDefinitions := make([]schema.MetricDefinition, 0)
+	for partition := partitionFrom; (partitionTo == -1 && partition == partitionFrom) || (partitionTo > 0 && partition < partitionTo); partition++ {
+		metricDefinitions = cassandraIndex.LoadPartitions([]int32{int32(partition)}, metricDefinitions, time.Now())
+	}
+
+	tracker := Tracker{}
+	for _, metricDefinition := range metricDefinitions {
+		tracker[metricDefinition.Id] = Track{
+			Name: metricDefinition.Name,
+			Tags: metricDefinition.Tags,
+		}
+	}
+
+	return &inputOOOFinder{
+		prefix: prefix,
+		substr: substr,
+
+		tracker:                  tracker,
+		reorderBufferGracePeriod: int64(metricInterval * reorderWindow),
+
+		groupByName:   groupByName,
+		groupedByName: groupedByName,
+		groupByTag:    groupByTag,
+		groupedByTag:  groupedByTag,
+
+		lock: sync.Mutex{},
+	}
+}
+
+func (ip *inputOOOFinder) processTrack(metricKey schema.MKey, metricTime int64, track Track) {
+	if ip.prefix != "" && !strings.HasPrefix(track.Name, ip.prefix) {
+		return
+	}
+	if ip.substr != "" && !strings.Contains(track.Name, ip.substr) {
+		return
+	}
+
+	if metricTime > track.LatestTimestamp {
+		track.LatestTimestamp = metricTime
+		ip.tracker[metricKey] = track
+	} else if metricTime+ip.reorderBufferGracePeriod < track.LatestTimestamp {
+		// increment grouping counts
+		if ip.groupByName == true {
+			(*ip.groupedByName)[track.Name]++
+		}
+
+		if ip.groupByTag != "" {
+			for _, tag := range track.Tags {
+				kv := strings.Split(tag, "=")
+				if len(kv) != 2 {
+					log.Errorf("unexpected tag encoding tag=%q", tag)
+					continue
+				}
+				if kv[0] == ip.groupByTag {
+					(*ip.groupedByTag)[kv[1]]++
+				}
+			}
+		}
+	}
+}
+
+func (ip *inputOOOFinder) ProcessMetricData(metric *schema.MetricData, partition int32) {
+	metricKey, err := schema.MKeyFromString(metric.Id)
+	if err != nil {
+		log.Errorf("failed to get metric key from id=%v: %s", metric.Id, err.Error())
+		return
+	}
+
+	ip.lock.Lock()
+	defer ip.lock.Unlock()
+
+	track, exists := ip.tracker[metricKey]
+	if !exists {
+		ip.tracker[metricKey] = Track{
+			Name:            metric.Name,
+			Tags:            metric.Tags,
+			LatestTimestamp: metric.Time,
+		}
+		return
+	}
+
+	ip.processTrack(metricKey, metric.Time, track)
+}
+
+func (ip *inputOOOFinder) ProcessMetricPoint(mp schema.MetricPoint, format msg.Format, partition int32) {
+	ip.lock.Lock()
+	defer ip.lock.Unlock()
+
+	track, exists := ip.tracker[mp.MKey]
+	if !exists {
+		log.Errorf("metric definition for key=%v not found", mp.MKey)
```

Comment:
I look forward to seeing what these are! We have a *very* low rate of them, but not zero

Reply:
We should also print the partition here, so we have an easier time of querying it out of the DB